### PR TITLE
Implement feature `integer_sign_cast`

### PR DESCRIPTION
**Pull request by Rua**
Tracking issue: https://github.com/rust-lang/rust/issues/125882

Since this is my first time making a library addition I wasn't sure where to place the new code relative to existing code. I decided to place it near the top where there are already some other basic bitwise manipulation functions. If there is an official guideline for the ordering of functions, please let me know.

### CODE REVIEW

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -183,6 +183,30 @@ macro_rules! int_impl {
             (self as $UnsignedT).trailing_ones()
         }
 
+        /// Returns the bit pattern of `self` reinterpreted as an unsigned integer of the same size.
+        ///
+        /// This is a bit safer than `as` because it wouldn't silently change the size if the code
```
Comment (Mark-Simulacrum) [inline library/core/src/num/int_macros.rs:188]:
```suggestion
        /// This produces the same result as an `as` cast, but ensures that the bit-width remains the same.
```

I don't think "safe" is a good term to use here given `as` casts are equally safe in the unsafe/safe sense. Something like this maybe?

On pointers a similar method has no extra explanation, so maybe just dropping this line is also fine: https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.cast-1

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with docs adjusted

The placement of the two functions seems fine, I don't think there's any hard rule on ordering here.

File: library/core/src/num/int_macros.rs (hunk 1 above)
Comment (Rua) [reply library/core/src/num/int_macros.rs:188]: I took the wording from the `cast_const` function of pointers, where it's also not really a safety thing is it? Maybe it ought to be changed there then.

**Pull request merged**